### PR TITLE
Support scanning multi-version episodes

### PR DESCRIPTION
**Pull request by AllSeeingEyeTolledEweSew**
**Changes**
This is an attempt to implement https://features.jellyfin.org/posts/251/multiple-versions-drop-down-on-tv-series

I added support for a directory structure like `Game of Thrones (2011)/Season 1/Episode S01E01/Episode S01E01 - v1.mkv`. The "episode directory" can hold multiple versions of the same episode, which get automatically picked up by the scanner.

I attempted to *not* change any other case.

I tested with the following directory structure:

```
./game of thrones (2011)/extras/top-level-foo-in-extras.strm
./game of thrones (2011)/otherrrr/top-extra-foo.strm
./game of thrones (2011)/Season 1/Episode s01e01.strm
./game of thrones (2011)/Season 1/Episode s01e01 - version2.strm
./game of thrones (2011)/Season 1/Episode s01e02.strm
./game of thrones (2011)/Season 1/Episode s01e02 - version2.strm
./game of thrones (2011)/Season 1/Episode s01e03/Episode s01e03.strm
./game of thrones (2011)/Season 1/Episode s01e03/Episode s01e03 - version2.strm
./game of thrones (2011)/Season 1/extras/extra-level-two/s1-extra-level-two.strm
./game of thrones (2011)/Season 1/extras/s1-in-extras.strm
./game of thrones (2011)/Season 1/extras/s1-in-extras - version2.strm
./game of thrones (2011)/Season 1/otherr/s1-extra-bar.strm
./game of thrones (2011)/Season 1/otherr/s1-extra-foo.strm
./game of thrones (2011)/Season 1/s1-outer-foo.strm
./game of thrones (2011)/top-level-foo.strm
```

I verified that in the above structure, only s01e03 gets resolved to a single item having multiple versions. All other files above are resolved in the same way as before.

### CODE REVIEW

**Comment by AllSeeingEyeTolledEweSew** [comment]
We could add support for a bunch of other stuff, like extras and photos belonging to the episode in the directory. However I thought it best to start with the basics

**Comment by crobibero** [comment]
Since you're changing the scanner, can you please add some tests to support the changes?

**Review by crobibero — APPROVED** [review]

**Comment by Artiume** [comment]
I'm not the most versed in this area but ideally, we shouldnt need to make a folder for each specific episode, could this be implemented in `IMultiItemResolver` and handle the multi-thing in the method `IMultiItemResolver.ResolveMultiple`

**Comment by AllSeeingEyeTolledEweSew** [comment]
> I'm not the most versed in this area but ideally, we shouldnt need to make a folder for each specific episode, could this be implemented in `IMultiItemResolver` and handle the multi-thing in the method `IMultiItemResolver.ResolveMultiple`

**TL;DR**: I agree, but the state of the code means any further work will be many times more complex than this PR. This PR solves the user story today for relatively little work, and (AFAICT) little technical debt.

Today, for e.g. a flat directory of movies, the `ResolveMultiple` implementation uses `VideoListResolver`. This is not set up to take a flat directory and produce multiple items in multiple *versions*. It can only take a flat directory and group *multi-part* videos (`foo-cd1.mkv`, `foo-cd.mkv`).

So to resolve multiple versions of multiple episodes in the same folder as you describe, we'd need to either duplicate logic from `VideoListResolver` or else change it, which would change movie resolving significantly. This also puts new pressure on regex logic to do the grouping correctly, which is daunting.

Also note that resolver interfaces today map *one* file-or-folder to zero or more `BaseItem`s. `MovieResolver`'s `ResolveMultiple` interface maps the whole movies *directory* to a list of `Movie`s. This is different than how tv shows work: `SeasonResolver` maps a season directory to a `Season` item, and children of the season directory get *separately*, *individually* resolved with `EpisodeResolver`. Neither of these patterns can resolve a season directory to a *hierarchy* of `Season` and `Episode`s, as in the scenario you describe. I believe doing this would require significant refactoring.

**Comment by Bitwolfies** [comment]
Does this fix the issue where you cannot group more than 2 episodes? First time I tried to do it it crashed the server once I looked at the show (which I had to remove to fix) And currently while it will group and show all episodes in the drop down, the third episode you add wont be hidden from the season view and will still count as its own episode.

**Comment by Artiume** [comment]
This PR won't change any behavior you're currently experiencing. It will enable another option. You can make subfolders for each episode in each season folder. So for a 26 episode series, you'd need 26 subfolders, one for each episode. I'm personally not in favor of this behavior, I don't want to have to make new folders for each episode.

**Comment by AllSeeingEyeTolledEweSew** [comment]
As a user, I manage my library's file layout with scripts. It makes no difference to me to create one directory structure over another, so I think this feature is valid as-is. I know it doesn't satisfy *all* users, but I assert it satisfies *some*.

I would say that I lack the experience to write a PR to change the resolver code any further.

From what I can see, it would take significant refactoring to support a more "elegant" layout of multi-version episodes (a flat layout of multiple episodes in multiple versions). A refactor ought to include a lot of tests for reviewers to trust it, but
* I lack csharp experience to know what testable components look like, and
* I lack Jellyfin experience to know what meaningful tests look like.

**Comment by cvium** [comment]
The multi handling is quite ridiculous and prone to breaking, so I fully understand your hesitation. I think your change is fairly small and self-contained, but I'm a bit torn. It's not conceptually how we'd like to handle multi-episode parsing.
